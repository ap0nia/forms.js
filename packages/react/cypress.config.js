// @ts-check

import { defineConfig } from 'cypress'

const config = defineConfig({
  e2e: {
    baseUrl: 'http://localhost:5173/',
  },
})

export default config
