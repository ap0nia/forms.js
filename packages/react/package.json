{
  "name": "@forms.js/react",
  "version": "0.0.0",
  "description": "form.js react",
  "private": true,
  "license": "MIT",
  "homepage": "https://ap0nia.github.io/forms.js",
  "funding": "https://github.com/ap0nia/forms.js",
  "contributors": [
    {
      "name": "ap0nia",
      "email": "",
      "url": "https://github.com/ap0nia"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/ap0nia/forms.js",
    "directory": "packages/react"
  },
  "bugs": {
    "url": "https://github.com/ap0nia/forms.js/issues"
  },
  "keywords": [],
  "type": "module",
  "scripts": {
    "test": "vitest run __tests__/use-form-state.test"
  },
  "dependencies": {
    "@forms.js/common": "workspace:^",
    "@forms.js/core": "workspace:^",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "use-sync-external-store": "^1.2.0"
  },
  "devDependencies": {
    "@testing-library/dom": "^9.3.3",
    "@testing-library/jest-dom": "^6.1.3",
    "@testing-library/react": "^14.0.0",
    "@types/react": "^18.2.24",
    "@types/react-dom": "^18.2.8",
    "@types/use-sync-external-store": "^0.0.4",
    "msw": "^1.3.2",
    "typescript": "^5.1.6"
  },
  "packageManager": "pnpm@8.6.12",
  "engines": {
    "pnpm": "^8.0.0",
    "node": ">=18"
  }
}
