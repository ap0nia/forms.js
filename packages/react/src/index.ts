export * from './control'
export * from './controller'
// export * from './form';
// export * from './logic';
// export * from './types';
export * from './use-controller'
export * from './use-field-array'
export * from './use-form'
export * from './use-form-context'
// export * from './useFormState';
// export * from './useWatch';
// export * from './utils';
