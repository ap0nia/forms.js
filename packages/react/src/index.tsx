import { createRoot } from 'react-dom/client'

import { App } from './App'

function main() {
  const id = 'root'

  const root = document.getElementById(id)

  if (root == null) {
    throw new Error(`Element with id ${id} not found`)
  }

  createRoot(root).render(<App />)
}

main()
