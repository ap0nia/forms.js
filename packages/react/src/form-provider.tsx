import { createContext, useCallback, useSyncExternalStore } from 'react'

import type { Control } from './control'
import type { UseFormReturn } from './use-form'

export type FormProviderProps<TValues extends Record<string, any>, TContext = any> = {
  children?: React.ReactNode
  control: Control<TValues, TContext>
} & Partial<UseFormReturn<TValues, TContext>>

export const FormContext = createContext<UseFormReturn<any, any>>(undefined!)

export function FormProvider<TValues extends Record<string, any>, TContext = any>(
  props: FormProviderProps<TValues, TContext>,
) {
  const { control, ...rest } = props

  const subscribe = useCallback(
    (callback: () => void) => control.state.subscribe(callback),
    [control],
  )

  const getSnapshot = useCallback(() => control.state.value, [control])

  const getServerSnapshot = useCallback(() => control.state.value, [control])

  useSyncExternalStore(subscribe, getSnapshot, getServerSnapshot)

  const value = {
    control,
    ...rest,
  } as UseFormReturn<TValues, TContext>

  return <FormContext.Provider value={value}>{props.children}</FormContext.Provider>
}
