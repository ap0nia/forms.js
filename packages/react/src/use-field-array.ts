import { FieldArray, type FieldArrayOptions } from '@forms.js/core'
import type { NestedObjectArrays } from '@forms.js/core/utils/types/nested-object-arrays'
import { useCallback, useEffect, useRef, useSyncExternalStore } from 'react'

import type { Control } from './form-control'
import { useFormContext } from './use-form-context'

export type UseFieldArrayProps<
  TValues extends Record<string, any>,
  TContext = any,
  TTransformedValues extends Record<string, any> | undefined = undefined,
  TFieldArray extends NestedObjectArrays<TValues> = NestedObjectArrays<TValues>,
  TFieldArrayName extends Extract<keyof TFieldArray, string> = Extract<keyof TFieldArray, string>,
> = Omit<
  FieldArrayOptions<TValues, TContext, TTransformedValues, TFieldArray, TFieldArrayName>,
  'control'
> & {
  control?: Control<TValues, TContext, TTransformedValues>
}

export function useFieldArray<
  TValues extends Record<string, any>,
  TContext = any,
  TTransformedValues extends Record<string, any> | undefined = undefined,
  TFieldArray extends NestedObjectArrays<TValues> = NestedObjectArrays<TValues>,
  TFieldArrayName extends Extract<keyof TFieldArray, string> = Extract<keyof TFieldArray, string>,
>(props: UseFieldArrayProps<TValues, TContext, TTransformedValues, TFieldArray, TFieldArrayName>) {
  const context = useFormContext<TValues, TContext, TTransformedValues>()

  const control = props.control ?? context.control

  const fieldArray = useRef(
    new FieldArray<TValues, TContext, TTransformedValues, TFieldArray, TFieldArrayName>({
      ...props,
      control,
    }),
  )

  const subscribe = useCallback(
    (callback: () => void) => {
      return fieldArray.current.fields.subscribe(callback)
    },
    [fieldArray.current],
  )

  const getSnapshot = useCallback(() => {
    return fieldArray.current.fields.value
  }, [fieldArray.current])

  const getServerSnapshot = useCallback(() => {
    return fieldArray.current.fields.value
  }, [fieldArray.current])

  const fields = useSyncExternalStore(subscribe, getSnapshot, getServerSnapshot)

  fieldArray.current.mount()

  useEffect(() => {
    return () => {
      fieldArray.current.unmount()
    }
  }, [fieldArray.current])

  return {
    fields,
    swap: fieldArray.current.swap.bind(fieldArray.current),
    move: fieldArray.current.move.bind(fieldArray.current),
    prepend: fieldArray.current.prepend.bind(fieldArray.current),
    append: fieldArray.current.append.bind(fieldArray.current),
    remove: fieldArray.current.remove.bind(fieldArray.current),
    insert: fieldArray.current.insert.bind(fieldArray.current),
    update: fieldArray.current.update.bind(fieldArray.current),
    replace: fieldArray.current.replace.bind(fieldArray.current),
  }
}
