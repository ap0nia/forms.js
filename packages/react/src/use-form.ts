import { useRef, useCallback, useSyncExternalStore, useEffect } from 'react'

import { ReactFormControl, type FormControlOptions } from './form-control'

export function useForm<TValues extends Record<string, any>, TContext = any>(
  options?: FormControlOptions<TValues, TContext>,
) {
  const formControlRef = useRef<ReactFormControl<TValues, TContext>>()

  formControlRef.current ??= new ReactFormControl(options)

  const formControl = formControlRef.current

  const form = useRef({
    register: formControl.registerReact.bind(formControl),
    handleSubmit: formControl.handleSubmitReact.bind(formControl),
    unregister: formControl.unregister.bind(formControl),
    formState: formControl.derivedState.proxy,
    watch: formControl.watch.bind(formControl),
    reset: formControl.reset.bind(formControl),
    setError: formControl.setError.bind(formControl),
    clearErrors: formControl.clearErrors.bind(formControl),
    setValue: formControl.setValue.bind(formControl),
    setFocus: formControl.setFocus.bind(formControl),
    getValues: formControl.getValues.bind(formControl),
    getFieldState: formControl.getFieldState.bind(formControl),
    trigger: formControl.trigger.bind(formControl),
    formControl,
  })

  const subscribe = useCallback((callback: () => void) => {
    return formControl.derivedState.subscribe(callback)
  }, [])

  const getSnapshot = useCallback(() => {
    return formControl.derivedState.value
  }, [])

  useSyncExternalStore(subscribe, getSnapshot, getSnapshot)

  useEffect(() => {
    if (formControl.state.status.value.mount) {
      formControl.reset(options?.values, formControl.options.resetOptions)
    }
  }, [options?.values])

  useEffect(() => {
    formControl.cleanup()
  })

  useEffect(() => {
    formControl.mount()

    return () => {
      formControl.unmount()
    }
  }, [])

  return form.current
}
