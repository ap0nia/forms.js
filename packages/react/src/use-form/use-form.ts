import { FormControl } from '@forms.js/core'
import type {
  FormControlState,
  FormControlValues,
  FormControlOptions,
  RegisterOptions,
} from '@forms.js/core'
import { getRuleValue } from '@forms.js/core/validation/get-rule-value'
import { useCallback, useEffect, useMemo, useRef, useState } from 'react'

export type UseFormRegister<TValues extends Record<string, any>> = <
  TKey extends keyof FormControlValues<TValues>,
>(
  name: Extract<TKey, string>,
  options?: RegisterOptions<TValues, TKey>,
) => RegisterProps<TKey>

export type RegisterProps<T> = {
  onChange: ChangeHandler
  onBlur: ChangeHandler
  ref: React.RefCallback<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement>
  name: T
  min?: string | number
  max?: string | number
  maxLength?: number
  minLength?: number
  pattern?: string
  required?: boolean
  disabled?: boolean
}

export type ChangeHandler = (event: React.ChangeEvent) => unknown

// export type UseFormReturn<TValues extends Record<string, any>, TContext = any> = {
//   formControl: FormControl<TValues, TContext>
//   register: UseFormRegister<TValues>
//   getValues: GetValues<TValues>
//   handleSubmit: HandleSubmit<TValues>
// }

export function useForm<TValues extends Record<string, any>, TContext = any>(
  options?: FormControlOptions<TValues, TContext>,
) {
  const formControl = useMemo(() => {
    return new FormControl<TValues, TContext>(options)
  }, [options])

  const [values, setValues] = useState(formControl.state.values.value)

  const register = useCallback<UseFormRegister<TValues>>(
    (name, options = {}) => {
      const { registerElement, unregisterElement } = formControl.register(name, options)

      const props: RegisterProps<typeof name> = {
        name,
        ref: (instance) => {
          if (instance) {
            registerElement(instance)
          } else {
            unregisterElement()
          }
        },
        onChange: (event) => formControl.handleChange(event.nativeEvent),
        onBlur: (event) => formControl.handleChange(event.nativeEvent),
      }

      if (options?.disabled) {
        props.disabled = options.disabled
      }

      if (formControl.options.progressive) {
        props.required = !!options.required
        props.min = getRuleValue(options.min)
        props.max = getRuleValue(options.max)
        props.minLength = getRuleValue(options.minLength) as number
        props.maxLength = getRuleValue(options.maxLength) as number
        props.pattern = getRuleValue(options.pattern) as string
      }

      return props
    },
    [formControl, options],
  )

  const setter = useCallback((newValues: any) => {
    setValues({ ...newValues })
    // setValues(() => ({ ...newValues }))
  },[])

  const formState = useRef<FormControlState<TValues>>({
    // isDirty: boolean,
    // isLoading: boolean,
    // isSubmitted: boolean,
    // isSubmitSuccessful: boolean,
    // isSubmitting: boolean,
    // isValidating: boolean,
    // isValid: boolean,
    // submitCount: number,
    // dirtyFields: {},
    // touchedFields: Partial<Readonly<DeepMap<T, boolean>>>,
    // defaultValues: DeepPartial<T>,
    // values: T,
    // errors: FieldErrors<T>,
    // status: FormControlStatus,
  } as any)

  useEffect(() => {
    formControl.state.values.subscribe(setter)

    return () => {
      // unsubscribeFunctions.current.forEach((unsubscribe) => unsubscribe())
    }
  }, [])

  return {
    formControl,
    register,
    getValues: formControl.getValues.bind(formControl),
    handleSubmit: formControl.handleSubmit.bind(formControl),
    formState: formState.current,
  }
}
