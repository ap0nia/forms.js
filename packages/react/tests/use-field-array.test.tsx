import { render, renderHook, fireEvent, screen } from '@testing-library/react'
import React from 'react'
import { describe, it, expect, beforeEach } from 'vitest'

import type { Control } from '../src/form-control'
import { FormProvider } from '../src/form-provider'
import { useFieldArray } from '../src/use-field-array'
import { useForm } from '../src/use-form'

let i = 0

function idGenerator() {
  return String(i++)
}

describe('useFieldArray', () => {
  beforeEach(() => {
    i = 0
  })

  describe('initialize', () => {
    it('should return default fields value', () => {
      const { result } = renderHook(() => {
        const { control } = useForm()
        return useFieldArray({
          control,
          name: 'test',
        })
      })

      expect(result.current.fields).toEqual([])
    })

    it('should populate default values into fields', () => {
      const { result } = renderHook(() => {
        const { control } = useForm({
          defaultValues: { test: [{ test: '1' }, { test: '2' }] },
        })
        return useFieldArray({
          idGenerator,
          control,
          name: 'test',
        })
      })

      expect(result.current.fields).toEqual([
        { test: '1', id: '0' },
        { test: '2', id: '1' },
      ])
    })

    it('should render with FormProvider', () => {
      const Provider = ({ children }: { children: React.ReactNode }) => {
        const methods = useForm()
        return <FormProvider {...methods}>{children}</FormProvider>
      }
      const { result } = renderHook(() => useFieldArray({ name: 'test' }), {
        wrapper: Provider,
      })
      expect(result.error).toBeUndefined()
    })
  })

  describe('with should unregister false', () => {
    it('should still remain input value with toggle', () => {
      const Component = () => {
        const { register, control } = useForm<{
          test: {
            value: string
          }[]
        }>()
        const [show, setShow] = React.useState(true)
        const { fields, append } = useFieldArray({
          control,
          name: 'test',
        })

        return (
          <form>
            {show &&
              fields.map((field, i) => (
                <input key={field.id} {...register(`test.${i}.value` as const)} />
              ))}
            <button type="button" onClick={() => append({ value: '' })}>
              append
            </button>
            <button type="button" onClick={() => setShow(!show)}>
              toggle
            </button>
          </form>
        )
      }

      render(<Component />)

      fireEvent.click(screen.getByRole('button', { name: 'append' }))
      expect(screen.getAllByRole('textbox').length).toEqual(1)
      fireEvent.click(screen.getByRole('button', { name: 'toggle' }))
      expect(screen.queryByRole('textbox')).not.toBeInTheDocument()
      fireEvent.click(screen.getByRole('button', { name: 'toggle' }))
      expect(screen.getAllByRole('textbox').length).toEqual(1)
    })

    it('should show errors during mount when mode is set to onChange', async () => {
      const Component = () => {
        const {
          register,
          control,
          formState: { isValid, errors },
        } = useForm<{ test: { value: string }[] }>({
          defaultValues: {
            test: [{ value: 'test' }],
          },
          resolver: async () => ({
            values: {},
            errors: {
              test: [{ value: { message: 'wrong', type: 'test' } }],
            },
          }),
          mode: 'onChange',
        })
        const { fields, append } = useFieldArray({ name: 'test', control })

        return (
          <form>
            {fields.map((field, i) => (
              <input key={field.id} {...register(`test.${i}.value` as const)} />
            ))}
            <button
              type="button"
              onClick={() =>
                append({
                  value: 'test',
                })
              }
            >
              append
            </button>

            {!isValid && <p>not valid</p>}
            {errors.test && <p>errors</p>}
          </form>
        )
      }

      render(<Component />)

      expect(await screen.findByRole('textbox')).toBeVisible()
      expect(await screen.findByText('not valid')).toBeVisible()
    })

    it('should retain input values during unmount', async () => {
      type FormValues = {
        test: { name: string }[]
      }

      const FieldArray = ({
        control,
        register,
      }: {
        control: Control<FormValues>
        register: UseFormRegister<FormValues>
      }) => {
        const { fields } = useFieldArray({
          control,
          name: 'test',
        })

        return (
          <div>
            {fields.map((item, index) => {
              return (
                <div key={item.id}>
                  <input {...register(`test.${index}.name`)} />
                </div>
              )
            })}
          </div>
        )
      }

      const App = () => {
        const [show, setShow] = React.useState(true)
        const { control, register } = useForm({
          shouldUnregister: false,
          defaultValues: {
            test: [{ name: 'test' }],
          },
        })

        return (
          <div>
            {show && <FieldArray control={control} register={register} />}
            <button type={'button'} onClick={() => setShow(!show)}>
              toggle
            </button>
          </div>
        )
      }

      render(<App />)

      fireEvent.change(screen.getByRole('textbox'), {
        target: { value: '12345' },
      })

      fireEvent.click(screen.getByRole('button'))

      fireEvent.click(screen.getByRole('button'))

      expect((screen.getByRole('textbox') as HTMLInputElement).value).toEqual('12345')
    })
  })

  describe('with resolver', () => {
    it('should provide updated form value each action', async () => {
      let formData = {}
      const Component = () => {
        const {
          register,
          control,
          formState: { isValid },
        } = useForm<{
          data: string
          test: { value: string }[]
        }>({
          resolver: (data) => {
            formData = data
            return {
              values: {},
              errors: {},
            }
          },
        })
        const { fields, append } = useFieldArray({ name: 'test', control })

        return (
          <div>
            <input {...register('data')} defaultValue="test" />
            {fields.map((field, i) => (
              <input key={field.id} {...register(`test.${i}.value` as const)} />
            ))}
            <button onClick={() => append({ value: '' })}>append</button>
            <span>{isValid && 'valid'}</span>
          </div>
        )
      }

      render(<Component />)

      expect(await screen.findByText('valid')).toBeVisible()

      fireEvent.click(screen.getByRole('button'))

      expect(formData).toEqual({
        data: 'test',
        test: [{ value: '' }],
      })
    })
  })

  it('should provide correct form data with nested field array', async () => {
    type FormValues = {
      test: {
        value: string
        nestedArray: {
          value: string
        }[]
      }[]
    }

    let formData: any = {}
    const Nested = ({ index, control }: { control: Control<FormValues>; index: number }) => {
      const { fields, append } = useFieldArray<FormValues>({
        name: `test.${index}.nestedArray` as const,
        control,
      })

      return (
        <div>
          {fields.map((item, i) => (
            <input
              key={item.id}
              {...control.registerReact(`test.${index}.nestedArray.${i}.value` as const)}
            />
          ))}

          <button type={'button'} onClick={() => append({ value: 'test' })}>
            Append Nest
          </button>
        </div>
      )
    }

    const Component = () => {
      const {
        register,
        control,
        formState: { isValid },
      } = useForm<FormValues>({
        resolver: (data) => {
          formData = data
          return {
            values: data,
            errors: {},
          }
        },
        mode: 'onChange',
        defaultValues: {
          test: [{ value: '1', nestedArray: [{ value: '2' }] }],
        },
      })
      const { fields, remove } = useFieldArray({
        name: 'test',
        control,
      })

      return (
        <form>
          {fields.map((item, i) => (
            <fieldset key={item.id}>
              <input {...register(`test.${i}.value` as const)} />

              <Nested control={control} index={i} />
              <button type={'button'} onClick={() => remove(i)}>
                delete
              </button>
            </fieldset>
          ))}
          <span>{isValid && 'valid'}</span>
        </form>
      )
    }

    render(<Component />)

    fireEvent.click(screen.getByRole('button', { name: 'Append Nest' }))

    expect(await screen.findByText('valid')).toBeVisible()

    expect(formData).toEqual({
      test: [
        {
          value: '1',
          nestedArray: [{ value: '2' }, { value: 'test' }],
        },
      ],
    })

    fireEvent.click(screen.getByRole('button', { name: 'delete' }))

    expect(formData).toEqual({
      test: [],
    })
  })

  it('should report field array error during user action', async () => {
    const App = () => {
      const {
        register,
        control,
        formState: { errors },
      } = useForm<{
        test: { value: string }[]
      }>({
        mode: 'onChange',
        resolver: (data) => {
          return {
            values: data,
            errors: {
              test: {
                type: 'test',
                message: 'minLength',
              },
            },
          }
        },
        defaultValues: {
          test: [{ value: '1' }],
        },
      })
      const { fields, remove } = useFieldArray({
        name: 'test',
        control,
      })

      return (
        <form>
          {errors.test && <p>minLength</p>}

          {fields.map((item, i) => (
            <fieldset key={item.id}>
              <input {...register(`test.${i}.value` as const)} />
              <button type={'button'} onClick={() => remove(i)}>
                delete
              </button>
            </fieldset>
          ))}
        </form>
      )
    }

    render(<App />)

    expect(screen.queryByText('minLength')).not.toBeInTheDocument()

    fireEvent.click(screen.getByRole('button'))

    expect(await screen.findByText('minLength')).toBeVisible()
  })
})
