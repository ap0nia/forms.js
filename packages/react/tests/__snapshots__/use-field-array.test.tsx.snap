// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFieldArray > submit form > should append multiple inputs correctly 1`] = `
[
  {
    "test": [
      {
        "value": "data",
      },
    ],
  },
  {
    "test": [
      {
        "value": "data",
      },
      {
        "value": "test",
      },
      {
        "value": "test1",
      },
    ],
  },
]
`;
